File: day06/python/part2.py
#!/usr/bin/env python3

import helper

from typing import Dict, List, Set


def find_path(d: Dict[str, str], src: str, dest: str) -> List[str]:
    path = []
    key = src
    while True:
        value = d[key]
        path.append(value)
        if value == dest:
            break
        key = value
    #
    return path[::-1]


def main() -> None:
    # lines = helper.read_lines("example2.txt")
    lines = helper.read_lines("input.txt")

    child_parent: Dict[str, str] = {}
    for line in lines:
        left, right = line.split(')')
        child_parent[right] = left

    p1 = find_path(child_parent, 'YOU', 'COM')
    # print(p1)
    p2 = find_path(child_parent, 'SAN', 'COM')
    # print(p2)

    p1_set: Set[str] = set(p1)
    p2_set: Set[str] = set(p2)
    diff1 = len(p1_set.difference(p2_set))
    diff2 = len(p2_set.difference(p1_set))
    result = diff1 + diff2
    print(result)

##############################################################################

if __name__ == "__main__":
    main()

File: day06/rust/src/bin/part2.rs
use day06::helper::read_input_file;

use fxhash::FxHashMap as HashMap;
use std::collections::HashSet;

// const INPUT_FILE: &str = "example2.txt";
const INPUT_FILE: &str = "input.txt";


fn find_path(d: &HashMap<&str, &str>, src: &str, dest: &str) -> Vec<String>
{
    let mut path = Vec::new();
    let mut key = src;
    loop {
        let value = d[key];
        path.push(value.to_string());
        if value == dest {
            break;
        }
        key = value;
    }

    path.reverse();
    path
}

fn main()
{
    let lines: Vec<String> = read_input_file(INPUT_FILE);

    let mut child_parent: HashMap<&str, &str> = HashMap::default();
    for line in lines.iter() {
        let parts: Vec<&str> = line.split(")").collect();
        let (left, right) = (parts[0], parts[1]);
        child_parent.insert(right, left);
    }
    let child_parent = child_parent;    // immutable

    let p1: Vec<String> = find_path(&child_parent, "YOU", "COM");
    // println!("{:?}", p1);
    let p2: Vec<String> = find_path(&child_parent, "SAN", "COM");
    // println!("{:?}", p2);

    let p1_set: HashSet<String> = p1.iter().cloned().collect();
    // println!("{:?}", p1_set);
    let p2_set: HashSet<String> = p2.iter().cloned().collect();
    // println!("{:?}", p2_set);
    let diff1: HashSet<_> = p1_set.difference(&p2_set).collect();
    let diff2: HashSet<_> = p2_set.difference(&p1_set).collect();
    let result = diff1.len() + diff2.len();
    println!("{}", result);
}
